modules/standard/Set.chpl

```diff
@@ -195,17 +195,25 @@ module Set {
 
       :arg other: A set to initialize this set with.
     */
-    proc init=(const ref other: set(?t, ?)) lifetime this < other {
-      this.eltType = t;
-      this.parSafe = other.parSafe;
+    proc init=(const ref other: set(?t, ?p)) lifetime this < other {
+      this.eltType = if this.type.eltType != ? then
+                        this.type.eltType else t;
+      this.parSafe = if this.type.parSafe != ? then
+                        this.type.parSafe else p;
       this.complete();
 
-      if !isCopyableType(eltType) then
-        compilerError('cannot initialize ' + this.type:string + ' from ' +
-                      other.type:string + ' because element type ' +
-                      eltType:string + ' is not copyable');
-
-      for elem in other do _addElem(elem);
+      // TODO: Relax this to allow if 'isCoercible(t, this.eltType)'?
+      if this.eltType != t {
+        compilerError('cannot initialize ', this.type:string, ' from ',
+                      other.type:string, ' due to element type ',
+                      'mismatch');
+      } else if !isCopyableType(eltType) {
+        compilerError('cannot initialize ', this.type:string, ' from ',
+                      other.type:string, ' because element type ',
+                      eltType:string, ' is not copyable');
```

Comment:
The resulting code mixes `eltType` and `this.eltType`. I'd stick with the former, otherwise I suggest switching the remaining occurrences of `eltType` to the latter.